File: tests/tamm/Test_IndexLoopNest.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest/doctest.h"

#include <iostream>
#include <tamm/index_loop_nest.hpp>

using namespace tamm;

/*template<typename T>
std::ostream&
operator << (std::ostream& os, const std::vector<T>& vec) {
  os<<"[";
  for(const auto& v: vec) {
    os<<v<<",";
  }
  os<<"]"<<std::endl;
  return os;
}*/

TEST_CASE("Zero-dimensional index loop nest with index bound constructor") {
  IndexLoopNest iln{};
  int           cnt = 0;
  REQUIRE(iln.begin() != iln.end());
  std::for_each(iln.begin(), iln.end(), [&](const auto&) { cnt += 1; });
  REQUIRE(cnt == 1);
}

TEST_CASE("Zero-dimensional index loop nest with list of arguments constructor") {
  IndexLoopNest iln{{}, {}, {}, {}};
  int           cnt = 0;
  REQUIRE(iln.begin() != iln.end());
  std::for_each(iln.begin(), iln.end(), [&](const auto&) { cnt += 1; });
  REQUIRE(cnt == 1);
}

TEST_CASE("One-dimensional index loop nest with index bound constructor") {
  IndexSpace      is{range(10)};
  TiledIndexSpace tis{is, 1};
  TiledIndexLabel i;
  std::tie(i) = tis.labels<1>("all");

  IndexLoopNest iln{i};
  unsigned      cnt = 0;
  for(auto itr = iln.begin(); itr != iln.end(); itr++, cnt++) {
    // std::cout<<"--"<<*itr<<" cnt="<<cnt<<std::endl;
    REQUIRE(*itr == IndexVector{cnt});
  }
  REQUIRE(cnt == 10);
}

TEST_CASE("One-dimensional index loop nest with list of arguments constructor") {
  IndexSpace      is{range(10)};
  TiledIndexSpace tis{is, 1};
  IndexLoopNest   iln{{tis}, {}, {}, {}};
  unsigned        cnt = 0;
  for(auto itr = iln.begin(); itr != iln.end(); itr++, cnt++) {
    // std::cout<<"--"<<it<<std::endl;
    REQUIRE(*itr == IndexVector{cnt});
  }
  REQUIRE(cnt == 10);
}

TEST_CASE("Two-dimensional square index loop nest with index bound constructor") {
  IndexSpace      is{range(10)};
  TiledIndexSpace tis{is, 1};
  TiledIndexLabel i, j;
  std::tie(i, j) = tis.labels<2>("all");
  IndexLoopNest iln{i, j};
  auto          itr = iln.begin();
  for(unsigned ci = 0; ci < 10; ci++) {
    for(unsigned cj = 0; cj < 10; cj++, itr++) {
      REQUIRE(itr != iln.end());
      REQUIRE(*itr == IndexVector{ci, cj});
    }
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Two-dimensional square index loop nest with list of arguments constructor") {
  IndexSpace      is{range(10)};
  TiledIndexSpace tis{is, 1};
  IndexLoopNest   iln{{tis, tis}, {}, {}, {}};
  auto            itr = iln.begin();
  for(unsigned ci = 0; ci < 10; ci++) {
    for(unsigned cj = 0; cj < 10; cj++, itr++) {
      REQUIRE(itr != iln.end());
      REQUIRE(*itr == IndexVector{ci, cj});
    }
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Two-dimensional rectangular  index loop nest") {
  const unsigned  ri1 = 9, ri2 = 23;
  IndexSpace      is1{range(ri1)}, is2{range(ri2)};
  TiledIndexSpace tis1{is1, 1}, tis2{is2, 1};
  TiledIndexLabel til1, til2;
  std::tie(til1) = tis1.labels<1>("all");
  std::tie(til2) = tis2.labels<1>("all");

  IndexLoopNest iln{til1, til2};
  auto          itr = iln.begin();
  for(unsigned c1 = 0; c1 < ri1; c1++) {
    for(unsigned c2 = 0; c2 < ri2; c2++, itr++) {
      REQUIRE(itr != iln.end());
      REQUIRE(*itr == IndexVector{c1, c2});
    }
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Two-dimensional upper triangular index loop nest") {
  const unsigned  ri = 11;
  IndexSpace      is{range(ri)};
  TiledIndexSpace tis{is, 1};
  TiledIndexLabel i, j;
  std::tie(i, j) = tis.labels<2>("all");
  IndexLoopNest iln{i, j + (IndexLoopBound{j} >= i)};
  auto          itr = iln.begin();
  for(unsigned ci = 0; ci < ri; ci++) {
    for(unsigned cj = ci; cj < ri; cj++, itr++) {
      REQUIRE(itr != iln.end());
      REQUIRE(*itr == IndexVector{ci, cj});
    }
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Two-dimensional lower triangular index loop nest") {
  const unsigned  ri = 11;
  IndexSpace      is{range(ri)};
  TiledIndexSpace tis{is, 1};
  TiledIndexLabel i, j;
  std::tie(i, j) = tis.labels<2>("all");
  IndexLoopNest iln{i, j + (IndexLoopBound{j} <= i)};
  auto          itr = iln.begin();
  for(unsigned ci = 0; ci < ri; ci++) {
    for(unsigned cj = 0; cj <= ci; cj++, itr++) {
      REQUIRE(itr != iln.end());
      REQUIRE(*itr == IndexVector{ci, cj});
    }
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Two-dimensional diagonal index loop nest") {
  const unsigned  ri = 11;
  IndexSpace      is{range(ri)};
  TiledIndexSpace tis{is, 1};
  TiledIndexLabel i, j;
  std::tie(i, j) = tis.labels<2>("all");
  IndexLoopNest iln{i, j + (IndexLoopBound{j} <= i) + (IndexLoopBound{j} >= i)};
  auto          itr = iln.begin();
  for(unsigned ci = 0; ci < ri; ci++, itr++) {
    REQUIRE(itr != iln.end());
    REQUIRE(*itr == IndexVector{ci, ci});
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Three-dimensional diagonal index loop nest") {
  const unsigned  ri = 11;
  IndexSpace      is{range(ri)};
  TiledIndexSpace tis{is, 1};
  TiledIndexLabel i, j, k;
  std::tie(i, j, k) = tis.labels<3>("all");
  IndexLoopNest iln{i, j + (IndexLoopBound{j} <= i) + (IndexLoopBound{j} >= i),
                    IndexLoopBound{k} == j};
  auto          itr = iln.begin();
  for(unsigned ci = 0; ci < ri; ci++, itr++) {
    REQUIRE(itr != iln.end());
    REQUIRE(*itr == IndexVector{ci, ci, ci});
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("One-dimensional split index loop nest with index bound constructor") {
  IndexSpace      is{range(20), {{"r1", {range(0, 10)}}, {"r2", {range(10, 20)}}}};
  TiledIndexSpace tis{is, 1};
  TiledIndexLabel i;
  std::tie(i) = tis.labels<1>("all");

  IndexLoopNest iln{i};
  unsigned      cnt = 0;
  for(auto itr = iln.begin(); itr != iln.end(); itr++, cnt++) {
    // std::cout<<"--"<<it<<std::endl;
    REQUIRE(*itr == IndexVector{cnt});
  }
  REQUIRE(cnt == 20);
}

TEST_CASE("Two-dimensional split index loop nest with index bound constructor") {
  IndexSpace      is{range(20), {{"r1", {range(0, 10)}}, {"r2", {range(10, 20)}}}};
  TiledIndexSpace tis{is, 10};
  TiledIndexLabel i, j;
  std::tie(i, j) = tis.labels<2>("all");
  IndexLoopNest iln{i, j};
  auto          itr = iln.begin();
  for(unsigned c1 = 0; c1 < 20 / 10; c1++) {
    for(unsigned c2 = 0; c2 < 20 / 10; c2++, itr++) {
      REQUIRE(itr != iln.end());
      REQUIRE(*itr == IndexVector{c1, c2});
    }
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Three-dimensional split index loop nest with index bound constructor") {
  Tile            tilesize = 10;
  IndexSpace      is{range(20), {{"r1", {range(0, 10)}}, {"r2", {range(10, 20)}}}};
  TiledIndexSpace tis{is, tilesize};
  TiledIndexLabel i, j, k;
  std::tie(i, j, k) = tis.labels<3>("all");
  IndexLoopNest iln{i, j, k};
  auto          itr = iln.begin();
  for(unsigned c1 = 0; c1 < 20 / tilesize; c1++) {
    for(unsigned c2 = 0; c2 < 20 / tilesize; c2++) {
      for(unsigned c3 = 0; c3 < 20 / tilesize; c3++, itr++) {
        REQUIRE(itr != iln.end());
        REQUIRE(*itr == IndexVector{c1, c2, c3});
      }
    }
  }
  REQUIRE(itr == iln.end());
}

TEST_CASE("Four-dimensional split index loop nest with index bound constructor") {
  Tile            tilesize = 10;
  IndexSpace      is{range(20), {{"r1", {range(0, 10)}}, {"r2", {range(10, 20)}}}};
  TiledIndexSpace tis{is, tilesize};
  TiledIndexLabel i, j, k, l;
  std::tie(i, j, k, l) = tis.labels<4>("all");
  IndexLoopNest iln{i, j, k, l};
  auto          itr = iln.begin();
  for(unsigned c1 = 0; c1 < 20 / tilesize; c1++) {
    for(unsigned c2 = 0; c2 < 20 / tilesize; c2++) {
      for(unsigned c3 = 0; c3 < 20 / tilesize; c3++) {
        for(unsigned c4 = 0; c4 < 20 / tilesize; c4++, itr++) {
          REQUIRE(itr != iln.end());
          REQUIRE(*itr == IndexVector{c1, c2, c3, c4});
        }
      }
    }
  }
  REQUIRE(itr == iln.end());
}

File: tests/python/Test_IndexLoopNest.py
import pytamm as pt


def assert_loop_values(iln, expected):
    got = [tuple(int(x) for x in v) for v in iln]
    expected = [tuple(int(x) for x in v) for v in expected]
    assert got == expected


def test_zero_dimensional_index_loop_nest_with_index_bound_constructor():
    iln = pt.IndexLoopNest()
    values = list(iln)
    assert len(values) == 1


def test_zero_dimensional_index_loop_nest_with_list_of_arguments_constructor():
    iln = pt.IndexLoopNest([])
    values = list(iln)
    assert len(values) == 1


def test_one_dimensional_index_loop_nest_with_index_bound_constructor():
    is0 = pt.IndexSpace(pt.range(10))
    tis = pt.TiledIndexSpace(is0, 1)
    i, = tis.labels("all", count=1)

    iln = pt.IndexLoopNest(i)

    expected = [(cnt,) for cnt in range(10)]
    assert_loop_values(iln, expected)


def test_one_dimensional_index_loop_nest_with_list_of_arguments_constructor():
    is0 = pt.IndexSpace(pt.range(10))
    tis = pt.TiledIndexSpace(is0, 1)

    iln = pt.IndexLoopNest([tis])

    expected = [(cnt,) for cnt in range(10)]
    assert_loop_values(iln, expected)


def test_two_dimensional_square_index_loop_nest_with_index_bound_constructor():
    is0 = pt.IndexSpace(pt.range(10))
    tis = pt.TiledIndexSpace(is0, 1)
    i, j = tis.labels("all", count=2)

    iln = pt.IndexLoopNest([i, j])

    expected = [(ci, cj) for ci in range(10) for cj in range(10)]
    assert_loop_values(iln, expected)


def test_two_dimensional_square_index_loop_nest_with_list_of_arguments_constructor():
    is0 = pt.IndexSpace(pt.range(10))
    tis = pt.TiledIndexSpace(is0, 1)

    iln = pt.IndexLoopNest([tis, tis])

    expected = [(ci, cj) for ci in range(10) for cj in range(10)]
    assert_loop_values(iln, expected)


def test_two_dimensional_rectangular_index_loop_nest():
    ri1 = 9
    ri2 = 23

    is1 = pt.IndexSpace(pt.range(ri1))
    is2 = pt.IndexSpace(pt.range(ri2))

    tis1 = pt.TiledIndexSpace(is1, 1)
    tis2 = pt.TiledIndexSpace(is2, 1)

    til1, = tis1.labels("all", count=1)
    til2, = tis2.labels("all", count=1)

    iln = pt.IndexLoopNest([til1, til2])

    expected = [(c1, c2) for c1 in range(ri1) for c2 in range(ri2)]
    assert_loop_values(iln, expected)


def test_two_dimensional_upper_triangular_index_loop_nest():
    ri = 11

    is0 = pt.IndexSpace(pt.range(ri))
    tis = pt.TiledIndexSpace(is0, 1)

    i, j = tis.labels("all", count=2)

    bi = pt.IndexLoopBound(i)
    bj = pt.IndexLoopBound(j)

    iln = pt.IndexLoopNest([bi, bj + (bj >= bi)])

    expected = [(ci, cj) for ci in range(ri) for cj in range(ci, ri)]
    assert_loop_values(iln, expected)


def test_two_dimensional_lower_triangular_index_loop_nest():
    ri = 11

    is0 = pt.IndexSpace(pt.range(ri))
    tis = pt.TiledIndexSpace(is0, 1)

    i, j = tis.labels("all", count=2)

    bi = pt.IndexLoopBound(i)
    bj = pt.IndexLoopBound(j)

    iln = pt.IndexLoopNest([bi, bj + (bj <= bi)])

    expected = [(ci, cj) for ci in range(ri) for cj in range(ci + 1)]
    assert_loop_values(iln, expected)


def test_two_dimensional_diagonal_index_loop_nest():
    ri = 11

    is0 = pt.IndexSpace(pt.range(ri))
    tis = pt.TiledIndexSpace(is0, 1)

    i, j = tis.labels("all", count=2)

    bi = pt.IndexLoopBound(i)
    bj = pt.IndexLoopBound(j)

    iln = pt.IndexLoopNest([bi, bj + (bj <= bi) + (bj >= bi)])

    expected = [(ci, ci) for ci in range(ri)]
    assert_loop_values(iln, expected)


def test_three_dimensional_diagonal_index_loop_nest():
    ri = 11

    is0 = pt.IndexSpace(pt.range(ri))
    tis = pt.TiledIndexSpace(is0, 1)

    i, j, k = tis.labels("all", count=3)

    bi = pt.IndexLoopBound(i)
    bj = pt.IndexLoopBound(j)
    bk = pt.IndexLoopBound(k)

    iln = pt.IndexLoopNest([bi, bj + (bj <= bi) + (bj >= bi), bk == bj])

    expected = [(ci, ci, ci) for ci in range(ri)]
    assert_loop_values(iln, expected)


def test_one_dimensional_split_index_loop_nest_with_index_bound_constructor():
    is0 = pt.IndexSpace(
        pt.range(20),
        {
            "r1": [pt.range(0, 10)],
            "r2": [pt.range(10, 20)],
        },
    )

    tis = pt.TiledIndexSpace(is0, 1)
    i, = tis.labels("all", count=1)

    iln = pt.IndexLoopNest(i)

    expected = [(cnt,) for cnt in range(20)]
    assert_loop_values(iln, expected)


def test_two_dimensional_split_index_loop_nest_with_index_bound_constructor():
    is0 = pt.IndexSpace(
        pt.range(20),
        {
            "r1": [pt.range(0, 10)],
            "r2": [pt.range(10, 20)],
        },
    )

    tis = pt.TiledIndexSpace(is0, 10)
    i, j = tis.labels("all", count=2)

    iln = pt.IndexLoopNest([i, j])

    expected = [(c1, c2) for c1 in range(2) for c2 in range(2)]
    assert_loop_values(iln, expected)


def test_three_dimensional_split_index_loop_nest_with_index_bound_constructor():
    tilesize = 10

    is0 = pt.IndexSpace(
        pt.range(20),
        {
            "r1": [pt.range(0, 10)],
            "r2": [pt.range(10, 20)],
        },
    )

    tis = pt.TiledIndexSpace(is0, tilesize)
    i, j, k = tis.labels("all", count=3)

    iln = pt.IndexLoopNest([i, j, k])

    expected = [
        (c1, c2, c3)
        for c1 in range(20 // tilesize)
        for c2 in range(20 // tilesize)
        for c3 in range(20 // tilesize)
    ]

    assert_loop_values(iln, expected)


def test_four_dimensional_split_index_loop_nest_with_index_bound_constructor():
    tilesize = 10

    is0 = pt.IndexSpace(
        pt.range(20),
        {
            "r1": [pt.range(0, 10)],
            "r2": [pt.range(10, 20)],
        },
    )

    tis = pt.TiledIndexSpace(is0, tilesize)
    i, j, k, l = tis.labels("all", count=4)

    iln = pt.IndexLoopNest([i, j, k, l])

    expected = [
        (c1, c2, c3, c4)
        for c1 in range(20 // tilesize)
        for c2 in range(20 // tilesize)
        for c3 in range(20 // tilesize)
        for c4 in range(20 // tilesize)
    ]

    assert_loop_values(iln, expected)


def main():
    tests = [
        test_zero_dimensional_index_loop_nest_with_index_bound_constructor,
        test_zero_dimensional_index_loop_nest_with_list_of_arguments_constructor,
        test_one_dimensional_index_loop_nest_with_index_bound_constructor,
        test_one_dimensional_index_loop_nest_with_list_of_arguments_constructor,
        test_two_dimensional_square_index_loop_nest_with_index_bound_constructor,
        test_two_dimensional_square_index_loop_nest_with_list_of_arguments_constructor,
        test_two_dimensional_rectangular_index_loop_nest,
        test_two_dimensional_upper_triangular_index_loop_nest,
        test_two_dimensional_lower_triangular_index_loop_nest,
        test_two_dimensional_diagonal_index_loop_nest,
        test_three_dimensional_diagonal_index_loop_nest,
        test_one_dimensional_split_index_loop_nest_with_index_bound_constructor,
        test_two_dimensional_split_index_loop_nest_with_index_bound_constructor,
        test_three_dimensional_split_index_loop_nest_with_index_bound_constructor,
        test_four_dimensional_split_index_loop_nest_with_index_bound_constructor,
    ]

    for test in tests:
        test()


if __name__ == "__main__":
    main()